docs/reference/components/sources/internal_metrics.cue

```diff
@@ -291,6 +291,66 @@ components: sources: internal_metrics: {
 				file: _file
 			}
 		}
+		kafka_queue_messages: {
+			description:       "Current number of messages in producer queues."
+			type:              "gauge"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
+		kafka_queue_messages_bytes: {
+			description:       "Current total size of messages in producer queues."
+			type:              "gauge"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
+		kafka_requests_total: {
+			description:       "Total number of requests sent to Kafka brokers."
+			type:              "gauge"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
+		kafka_requests_bytes_total: {
+			description:       "Total number of bytes transmitted to Kafka brokers."
+			type:              "gauge"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
+		kafka_responses_total: {
+			description:       "Total number of responses received from Kafka brokers."
+			type:              "gauge"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
+		kafka_responses_bytes_total: {
+			description:       "Total number of bytes received from Kafka brokers."
+			type:              "gauge"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
+		kafka_produced_messages_total: {
+			description:       "Total number of messages transmitted (produced) to Kafka brokers."
+			type:              "gauge"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
+		kafka_produced_messages_bytes_total: {
+			description:       "Total number of message bytes (including framing, such as per-Message framing and MessageSet/batch framing) transmitted to Kafka brokers."
+			type:              "gauge"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
+		kafka_consumed_messages_total: {
+			description:       "Total number of messages consumed, not including ignored messages (due to offset, etc), from Kafka brokers.."
+			type:              "gauge"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
+		kafka_consumed_messages_bytes_total: {
+			description:       "Total number of message bytes (including framing) received from Kafka brokers."
+			type:              "gauge"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
```

Comment:
These should all be counters, not gauges.

Reply:
Ah, I thought `counter`s only emit the delta since the last count and `gauge`s the absolute value, but apparently this interpretation is wrong.

Then I don't understand how all the `counter!("···", 1);` invocations can be correct in out codebase, e.g. [here](https://github.com/timberio/vector/blob/594a6c43e4e0f2d4e01136618f85a3a782eeec14/src/internal_events/kafka.rs#L15).

Reply:
…and here is where it gets a bit tricky.

Gauge metrics represent a point-in-time measurement of a value that may both increase and decrease. This fits things like processor utilization, memory size, instantaneous bandwidth use, etc. The `gauge!` macro emits a metric with this instantaneous value as an absolute metric.

Counter metrics represent an accumulator and never decrease (except on system restart). This fits things like network bytes sent, CPU time spent idle, etc. This is why all counter metrics are named `_total`. The `counter!` macro emits a metric with the amount to add to the total as an incremental metric.

The trick here is that you have the total values and need to emit absolute counter metrics, which can't be done directly with either the `gauge!` or `counter!` macros. To make matters worse, those are proc macros.

Reply:
Opened an issue in the `metrics` crate https://github.com/metrics-rs/metrics/issues/172 to see how to best proceed here.

Reply:
We could introduce a `update_counter!` macro in our codebase to calculate the delta before emitting the event here.

Reply:
Additional reference: https://github.com/timberio/vector/issues/6412. We need `metrics 0.14.x` to be able to use `layers::AbsoluteLayer` from `metrics-util 0.6.1`.

Reply:
Circling way back to this to note that I'm about to merge https://github.com/timberio/vector/pull/7031 which bumps to 0.14 with `layers::AbsoluteLayer` support.